WebDriverAgentLib/Categories/XCUIDevice+FBHelpers.m

```diff
@@ -37,13 +40,48 @@ - (BOOL)fb_goToHomescreenWithError:(NSError **)error
   return YES;
 }
 
-- (NSData *)fb_screenshot
+- (NSData *)fb_screenshotWithError:(NSError*__autoreleasing*)error
 {
   id xcScreen = NSClassFromString(@"XCUIScreen");
-  if (xcScreen) {
-    return (NSData *)[xcScreen valueForKeyPath:@"mainScreen.screenshot.PNGRepresentation"];
+  if (nil == xcScreen) {
+    NSData *result = [[XCAXClient_iOS sharedClient] screenshotData];
+    if (nil == result) {
+      if (error) {
+        *error = [[FBErrorBuilder.builder withDescription:@"Cannot take a screenshot of the current screen state"] build];
+      }
+      return nil;
+    }
+    return result;
+  }
+  
+  id mainScreen = [xcScreen valueForKey:@"mainScreen"];
+  CGSize screenSize = FBAdjustDimensionsForApplication(FBApplication.fb_activeApplication.frame.size, (UIInterfaceOrientation)[self.class sharedDevice].orientation);
+  SEL mSelector = NSSelectorFromString(@"screenshotDataForQuality:rect:error:");
+  NSMethodSignature *mSignature = [mainScreen methodSignatureForSelector:mSelector];
+  NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:mSignature];
```

Comment:
Why `NSInvocation`? Can't you simply add method definition in Private headers?

Reply:
We don't have XCUIScreen class present in private headers, since it was only introduced by Xcode9.

Reply:
Let me tweak, this and you can rebase your diff on my changes. It would make your life easier.

Reply:
Or I can refactor your code :)

Reply:
No problems. Thank you a lot for dedicating your time to review this stuff.